### ISSUE TITLE
Implement PyTorch Dataset and DataLoader for Steel Casting Sequences

### ISSUE THREAD

**Issue by dhar174** [dhar174/steel_defect_demo#68]
# PyTorch Dataset and DataLoader Implementation for Steel Casting Sequences

## Overview
Implement a comprehensive PyTorch Dataset and DataLoader system specifically designed for steel casting time series data. This implementation will serve as the data pipeline foundation for the LSTM model training and inference, providing efficient batch loading, memory optimization, and data augmentation capabilities.

### Primary Implementation
- **File**: `src/data/sequence_dataset.py`
- **Dependencies**: `torch>=1.13.0`, `numpy>=1.23.0`, `pandas>=1.5.0`, `scikit-learn>=1.1.0`
- **Integration**: Direct integration with LSTM model training pipeline and feature extraction system

#### 1. Custom PyTorch Dataset Class
````python
class CastingSequenceDataset(Dataset):
    """PyTorch Dataset for steel casting time series sequences"""
    
    def __init__(self, sequences: np.ndarray, labels: np.ndarray, 
                 transform=None, augment=False):
        """
        Initialize dataset with casting sequences and defect labels
        
        Parameters:
        - sequences: (N, sequence_length, num_features) array of sensor data
        - labels: (N,) array of binary defect labels
        - transform: Optional preprocessing transforms
        - augment: Enable data augmentation during training
Key Features Required:

Support for variable-length sequences with automatic padding/truncation
Memory-efficient data loading using lazy loading techniques
Integration with sequence preprocessing pipeline from src/features/feature_extractor.py
Configurable data transforms (normalization, augmentation)
Proper tensor conversion and device management
2. Efficient Batch Loading System
Python
def create_data_loaders(dataset_train, dataset_val, config):
    """
    Create optimized DataLoaders for training and validation
    
    Returns:
    - train_loader: DataLoader with shuffling and augmentation
    - val_loader: DataLoader for validation (no augmentation)
    """
Batch Loading Requirements:

Configurable batch sizes (default: 32 as per model config)
Automatic batching with proper tensor stacking
Shuffle enabled for training, disabled for validation
Pin memory optimization for GPU acceleration
Multi-worker support for parallel data loading
Proper handling of last incomplete batch
3. Memory-Efficient Data Management
Memory Optimization Features:

Lazy loading of sequence data to minimize RAM usage
Efficient tensor memory management with proper cleanup
Support for datasets larger than available RAM
Optimized data type usage (float32 vs float64)
Memory mapping for large datasets stored on disk
Garbage collection optimization during training
Implementation Strategy:

Python
class MemoryEfficientSequenceDataset(Dataset):
    def __init__(self, data_path, sequence_indices, labels):
        """Load only metadata, defer actual data loading to __getitem__"""
        self.data_path = data_path
        self.sequence_indices = sequence_indices
        self.labels = torch.FloatTensor(labels)
        
    def __getitem__(self, idx):
        """Load sequence data on-demand"""
        # Load only the required sequence from disk/memory
        pass
4. Weighted Sampling for Class Imbalance
Python
class ImbalancedSequenceSampler(WeightedRandomSampler):
    """Custom sampler to handle class imbalance in defect detection"""
    
    def __init__(self, labels, defect_weight_multiplier=3.0):
        """
        Create weighted sampler based on defect frequency
        
        Parameters:
        - labels: Binary defect labels
        - defect_weight_multiplier: Weight multiplier for defect samples
        """
Class Imbalance Handling:

Automatic weight calculation based on label distribution
Configurable weight multiplier for defect samples (default: 3.0)
Integration with PyTorch DataLoader sampler parameter
Support for stratified sampling during validation splits
Monitoring and logging of sample distribution per batch
5. Data Augmentation Integration
Python
class SequenceAugmentation:
    """Data augmentation techniques for time series sequences"""
    
    def __init__(self, config):
        self.noise_std = config.get('noise_std', 0.01)
        self.time_warp_sigma = config.get('time_warp_sigma', 0.2)
        self.magnitude_warp_sigma = config.get('magnitude_warp_sigma', 0.2)
    
    def add_noise(self, sequence):
        """Add Gaussian noise to sensor readings"""
        pass
    
    def time_warp(self, sequence):
        """Apply time warping augmentation"""
        pass
    
    def magnitude_warp(self, sequence):
        """Apply magnitude warping augmentation"""
        pass
Augmentation Techniques:

Gaussian noise injection (configurable standard deviation)
Time warping for temporal pattern variation
Magnitude warping for sensor reading variation
Random masking of sensor channels
Temporal shifting and scaling
Conditional augmentation (only on non-defect sequences)
Configuration Integration
Dataset Configuration (from configs/model_config.yaml)
YAML
lstm_model:
  data_processing:
    sequence_length: 300  # 5 minutes at 1Hz
    normalization: "z_score"
    padding: "zero"
    batch_size: 32
    num_workers: 4
    pin_memory: true
    
  augmentation:
    enabled: true
    noise_std: 0.01
    time_warp_probability: 0.3
    magnitude_warp_probability: 0.3
    
  sampling:
    weighted_sampling: true
    defect_weight_multiplier: 3.0
    stratify_validation: true
Implementation Details
File Structure
Code
src/data/sequence_dataset.py
├── CastingSequenceDataset class
├── MemoryEfficientSequenceDataset class
├── SequenceAugmentation class
├── ImbalancedSequenceSampler class
├── create_data_loaders function
├── sequence_collate_fn function
└── utilities for data validation
Integration Points
Feature Extraction: Interface with src/features/feature_extractor.py for preprocessing
Model Training: Direct integration with src/models/model_trainer.py
Configuration: Load settings from configs/model_config.yaml
Data Loading: Interface with synthetic data from data/synthetic/ directory
Custom Collate Function
Python
def sequence_collate_fn(batch):
    """
    Custom collate function for variable-length sequences
    
    Handles:
    - Padding sequences to max length in batch
    - Creating attention masks for padded sequences
    - Proper tensor stacking and device placement
    """
Performance Requirements
Speed Targets
Data Loading: < 50ms per batch on average
Memory Usage: < 2GB RAM for datasets up to 10,000 sequences
GPU Transfer: < 10ms per batch transfer to GPU
Augmentation: < 5ms per sequence for all augmentations combined
Scalability Requirements
Support datasets with up to 50,000 casting sequences
Handle sequence lengths from 60 to 600 time steps
Support 5-10 sensor features per time step
Efficient operation with batch sizes from 8 to 128
Testing Requirements
Unit Tests (tests/test_sequence_dataset.py)
Python
class TestCastingSequenceDataset:
    def test_dataset_initialization(self):
        """Test dataset creation with various input formats"""
        
    def test_sequence_padding_truncation(self):
        """Test handling of variable-length sequences"""
        
    def test_data_augmentation(self):
        """Test all augmentation techniques"""
        
    def test_weighted_sampling(self):
        """Test class imbalance handling"""
        
    def test_memory_efficiency(self):
        """Test memory usage with large datasets"""
        
    def test_dataloader_integration(self):
        """Test DataLoader creation and batch loading"""
Integration Tests
End-to-end data loading pipeline test
GPU acceleration compatibility test
Multi-worker data loading stability test
Memory leak detection during extended training
Compatibility with LSTM model input requirements
Deliverables
Core Implementation
CastingSequenceDataset class with full feature set
MemoryEfficientSequenceDataset for large datasets
SequenceAugmentation class with all augmentation techniques
ImbalancedSequenceSampler for class balance handling
create_data_loaders utility function
Custom collate function for variable-length sequences
Supporting Components
Comprehensive unit test suite (tests/test_sequence_dataset.py)
Integration test with LSTM training pipeline
Performance benchmarking script
Configuration validation utilities
Data pipeline documentation
Memory profiling and optimization report
Validation Criteria
Functional Validation
 Dataset correctly loads and processes synthetic casting sequences
 Batch loading produces consistent tensor shapes and types
 Data augmentation generates realistic variations without corrupting patterns
 Weighted sampling properly balances defect/non-defect samples
 Memory usage stays within specified limits for large datasets
Performance Validation
 Data loading speed meets performance targets (< 50ms per batch)
 GPU memory transfer is optimized and efficient
 Multi-worker data loading improves throughput without stability issues
 Memory usage scales linearly with dataset size
 No memory leaks during extended training sessions
Integration Validation
 Seamless integration with LSTM model training pipeline
 Proper tensor shapes and dtypes for model input requirements
 Configuration system correctly controls all dataset parameters
 Augmentation techniques improve model generalization without overfitting
 Class imbalance handling maintains training stability
Usage Examples
Basic Usage
Python
# Load and prepare data
from src.data.sequence_dataset import CastingSequenceDataset, create_data_loaders
from src.features.feature_extractor import SequenceFeatureExtractor

# Prepare sequences
extractor = SequenceFeatureExtractor(sequence_length=300)
sequences, labels = extractor.prepare_sequences(raw_data)

# Create datasets
train_dataset = CastingSequenceDataset(
    sequences=train_sequences, 
    labels=train_labels, 
    augment=True
)
val_dataset = CastingSequenceDataset(
    sequences=val_sequences, 
    labels=val_labels, 
    augment=False
)

# Create data loaders
train_loader, val_loader = create_data_loaders(
    train_dataset, val_dataset, config
)
Advanced Usage with Memory Efficiency
Python
# For large datasets
train_dataset = MemoryEfficientSequenceDataset(
    data_path='data/processed/sequences.h5',
    sequence_indices=train_indices,
    labels=train_labels
)

# With custom sampling
sampler = ImbalancedSequenceSampler(
    labels=train_labels,
    defect_weight_multiplier=4.0
)

train_loader = DataLoader(
    train_dataset,
    batch_size=32,
    sampler=sampler,
    collate_fn=sequence_collate_fn,
    num_workers=4,
    pin_memory=True
)
Dependencies and Environment
Required Packages
Python
torch>=1.13.0
numpy>=1.23.0
pandas>=1.5.0
scikit-learn>=1.1.0
h5py>=3.7.0  # For efficient data storage
tqdm>=4.64.0  # For progress tracking
Development Tools
Python
pytest>=7.2.0
pytest-cov>=4.0.0
memory-profiler>=0.60.0  # For memory usage analysis
line-profiler>=4.0.0     # For performance profiling
Implementation Priority
Phase 1: Core Implementation (Week 1)
 Basic CastingSequenceDataset class
 Standard DataLoader integration
 Basic sequence padding/truncation
 Simple unit tests
Phase 2: Optimization (Week 1-2)
 Memory-efficient dataset implementation
 Weighted sampling for class imbalance
 Multi-worker support optimization
 Performance benchmarking
Phase 3: Advanced Features (Week 2)
 Data augmentation integration
 Custom collate function
 Comprehensive testing suite
 Integration with LSTM training pipeline
This implementation forms a critical component of the LSTM model development phase, providing the robust data pipeline necessary for effective deep learning model training on steel casting time series data.
        """
